### PR TITLE
MINIFI-107 - Process group support

### PR DESCRIPTION
**Pull request by brosander**
Thank you for submitting a contribution to Apache NiFi - MiNiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi-minifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] - N/A - If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] - N/A - If applicable, have you updated the LICENSE file, including the main LICENSE file under minifi-assembly?
- [x] - N/A - If applicable, have you updated the NOTICE file, including the main NOTICE file found under minifi-assembly?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by JPercivall** [comment]
Reviewing

**Comment by JPercivall** [comment]
Also should probably update the default config.yml to include empty values for the new keys.

File: minifi-docs/src/main/markdown/System_Admin_Guide.md [hunk 1]
```diff
@@ -391,6 +391,39 @@ Within the Processor Configuration section, there is the `Properties` subsection
...
+name                                | The name of what this process group will do.
+id                                  | The id of this process group.  This needs to be set to a unique filesystem-friendly value (regex: [A-Za-z0-9_-]+)
+Processors                          | The processors contained in this Process Group. (Defined above)
+Remote Processing Groups            | The remote processing groups contained in this Process Group. (Defined below)
+Connections                         | The connections contained in this Process Group. (Defined below)
+Input Ports                         | The input ports contained in this Process Group. (Defined below)
+Output Ports                        | The output ports contained in this Process Group. (Defined below)
+Process Groups                      | The child Process Groups contained in this Process Group.
+
+## Input Ports
```
Comment (brosander) [inline minifi-docs/src/main/markdown/System_Admin_Guide.md]: Good point, adding clarifying text.  Also wrote up improvement [MINIFI-133](https://issues.apache.org/jira/browse/MINIFI-133).

File: minifi-docs/src/main/markdown/System_Admin_Guide.md [hunk 2]
```diff
@@ -391,6 +391,39 @@ Within the Processor Configuration section, there is the `Properties` subsection
...
+
+These ports provide input to the Process Group they reside on.
+
+*Property*               | *Description*
+--------------------     | -------------
+name                     | The name of what this input port will do.
+id                       | The id of this input port.  This needs to be set to a unique filesystem-friendly value (regex: [A-Za-z0-9_-]+)
+
+## Input Ports
```
Comment (brosander) [inline minifi-docs/src/main/markdown/System_Admin_Guide.md]: Copy pasta, fixing

File: minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/ConfigSchema.java [hunk 3]
```diff
@@ -101,47 +89,60 @@ public ConfigSchema(Map map, List<String> validationIssues) {
...
                 .flatMap(r -> r.getInputPorts().stream()).map(RemoteInputPortSchema::getId).collect(Collectors.toList());
-        checkForDuplicates(this::addValidationIssue, FOUND_THE_FOLLOWING_DUPLICATE_REMOTE_INPUT_PORT_IDS, remoteInputPortIdList);
-        remoteInputPortIds.addAll(remoteInputPortIdList);
+        List<String> allInputPortIds = allProcessGroups.stream().flatMap(p -> p.getInputPortSchemas().stream()).map(PortSchema::getId).collect(Collectors.toList());
+        List<String> allOutputPortIds = allProcessGroups.stream().flatMap(p -> p.getOutputPortSchemas().stream()).map(PortSchema::getId).collect(Collectors.toList());
+
+        checkForDuplicates(this::addValidationIssue, FOUND_THE_FOLLOWING_DUPLICATE_PROCESSOR_IDS, allProcessorIds);
+        checkForDuplicates(this::addValidationIssue, FOUND_THE_FOLLOWING_DUPLICATE_CONNECTION_IDS, allConnectionIds);
+        checkForDuplicates(this::addValidationIssue, FOUND_THE_FOLLOWING_DUPLICATE_REMOTE_PROCESSING_GROUP_NAMES, allRemoteProcessingGroupNames);
+        checkForDuplicates(this::addValidationIssue, FOUND_THE_FOLLOWING_DUPLICATE_REMOTE_INPUT_PORT_IDS, allRemoteInputPortIds);
```
Comment (brosander) [inline minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/ConfigSchema.java]: Adding checks for input as well as output (that check is only for remote input).

File: minifi-toolkit/minifi-toolkit-configuration/src/test/resources/ProcessGroupsAndRemoteProcessGroups.yml [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    name: null//UpdateAttribute
+    source id: 2078c936-0158-1000-0000-000000000000
+    source relationship names: []
+    destination id: 2078f34e-0158-1000-0000-000000000000
+    max work queue size: 10000
+    max work queue data size: 1 GB
+    flowfile expiration: 0 sec
+    queue prioritizer class: ''
+  - id: 207b1880-0158-1000-0000-000000000000
+    name: null//null
```
Comment (brosander) [inline minifi-toolkit/minifi-toolkit-configuration/src/test/resources/ProcessGroupsAndRemoteProcessGroups.yml]: Good catch, fixing the enrichment to recursively go through all flow snippets and to keep port names and ids in the map.

File: minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/ConfigSchema.java [hunk 5]
```diff
@@ -101,47 +89,60 @@ public ConfigSchema(Map map, List<String> validationIssues) {
...
-            addValidationIssue(FOUND_THE_FOLLOWING_DUPLICATE_IDS + duplicateIds.stream().sorted().collect(Collectors.joining(", ")));
+        OverlapResults<String> overlapResults = findOverlap(new HashSet<>(allProcessorIds), new HashSet<>(allRemoteInputPortIds), new HashSet<>(allInputPortIds), new HashSet<>(allOutputPortIds));
+        if (overlapResults.duplicates.size() > 0) {
+            addValidationIssue(FOUND_THE_FOLLOWING_DUPLICATE_IDS + overlapResults.duplicates.stream().sorted().collect(Collectors.joining(", ")));
         }
 
-        Set<String> connectableIds = new HashSet<>(processorIds);
-        connectableIds.addAll(remoteInputPortIds);
-        connections.forEach(c -> {
+        allConnectionSchemas.forEach(c -> {
```
Comment (brosander) [inline minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/ConfigSchema.java]: Correct, overlapResults.seen contains all potential connection sources and destinations.  Added a comment to that effect.

File: minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/ConfigSchema.java [hunk 6]
```diff
@@ -31,41 +33,33 @@
...
-    public static final String FOUND_THE_FOLLOWING_DUPLICATE_PROCESSOR_IDS = "Found the following duplicate processor ids: ";
-    public static final String FOUND_THE_FOLLOWING_DUPLICATE_CONNECTION_IDS = "Found the following duplicate connection ids: ";
-    public static final String FOUND_THE_FOLLOWING_DUPLICATE_REMOTE_PROCESSING_GROUP_NAMES = "Found the following duplicate remote processing group names: ";
-    public static final String FOUND_THE_FOLLOWING_DUPLICATE_REMOTE_INPUT_PORT_IDS = "Found the following duplicate remote input port ids: ";
-    public static final String FOUND_THE_FOLLOWING_DUPLICATE_IDS = "Found the following ids that occur both in Processors and Remote Input Ports: ";
     public static final int CONFIG_VERSION = 2;
     public static final String VERSION = "MiNiFi Config Version";
+    public static final String FOUND_THE_FOLLOWING_DUPLICATE_REMOTE_INPUT_PORT_IDS = "Found the following duplicate remote input port ids: ";
+    public static final String FOUND_THE_FOLLOWING_DUPLICATE_IDS = "Found the following ids that occur both in Processors and Remote Input Ports: ";
```
Comment (brosander) [inline minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/ConfigSchema.java]: Updating wording.

File: minifi-docs/src/main/markdown/System_Admin_Guide.md (hunk 1 above)
Comment (JPercivall) [inline minifi-docs/src/main/markdown/System_Admin_Guide.md]: Similar to what we had with PGs, during validation we should check if the root PG has input/output ports (and fail if so)

**Comment by brosander** [comment]
@JPercivall added validation that no input/output ports exist at top level

**Comment by JPercivall** [comment]
+1

Visually verified cod and did a contrib check build.  Ran a number of tests with templates -> config.yml verifying cascading PGs are correctly working (as well as running the configs) and all worked as expected. Thanks @brosander for bringing us one large step closer to feature parity with NiFi. I will merge it in.

**Pull request closed without merging**